lib/xapi-stdext-date/date.ml

```diff
@@ -12,18 +12,29 @@
  * GNU Lesser General Public License for more details.
  *)
 
-type iso8601 = string
+(** this dash type is used to store information about whether
+    the input string contained dashes or not (e.g. it is false when
+    `of_float` is called)
+  * we remember the dashes only where possible:
+     - `"yyyy-mm-ddTxx" |> of_string |> to_string = "yyyy-mm-ddTxx"
+  *  - `"yyyy-mm-ddTxx" |> of_string |> to_float |> of_float |> to_string = yyyymmddTxx`
+  *)
+type dash = bool
+
+type iso8601 = (string * dash)
```

Comment:
Since this time is abstract, can't we replace it with a Ptime, so that we know we have parsed it correctly?
The code below does a lot of juggling with offsets between UTC and localtime, where I think ptime could parse it independently of timezone. For backwards compatibility we could convert something without dashes into something that Ptime expects.

If we want to retain backwards compatibility wrt to where we raise exceptions you could have:
```
type iso8601 = {
   ptime:(Ptime.t * Ptime.t tz_offset_s option, `Msg of string) Result.t;
   orig: string;
   compat: bool
}
```

This would allow you to carry the date unchanged across of_string/to_string without raising exceptions (like the old code did), and raise exceptions only at the time of conversion to float, or at the assert utc.

Reply:
that sounds like a better idea :) i will implement that